hera_cal/redcal.py

```diff
@@ -288,6 +288,85 @@ def _build_polarity_baseline_groups(dly_cal_data, reds, edge_cut=0, max_rel_angl
             polarity_groups[red[0]] = (grp1, grp2)
 
     return polarity_groups
+
+
+def _determine_polarity_flips(polarity_groups, even_vs_odd_assumptions):
+    '''Take a set of polarity groups built by _build_polarity_baseline_groups() and set of a priori
+    assumptions about which ones are "even" (0 or 2 polarity flips) and which ones are "odd" (1 
+    polarity flip). Pick an antenna that participates mostly in even groups as a "not flipped"
+    reference. Use the assumptions to draw conclusions about more antennas. Use those anntennas
+    to draw conclusions about more groups being even/odd or odd/even. Keep iterating until it
+    finishes or gets stuck. If it fails to find a solution or reaches a contradiction, it raises
+    an AssertionError. Otherwise, it returns two dictionaries: is_flipped maps antennas to booleans
+    and even_vs_odd_IDs maps unique baseline keys to strings either 'even/odd' or 'odd/even'.
+    '''
+    even_vs_odd_IDs = deepcopy(even_vs_odd_assumptions)
+
+    # Find all antennas in the polarity groups
+    ants = set([ant for grp_pair in polarity_groups.values() 
+                for grp in grp_pair for bl in grp for ant in utils.split_bl(bl)])
+
+    # Count how many times each antenna is involved in an assumed even group minus odd group
+    ant_even_counts = {ant: 0 for ant in ants}
+    for key, (grp1, grp2) in polarity_groups.items():
+        if key in even_vs_odd_IDs:
+            even, odd = {'even/odd': (grp1, grp2), 'odd/even': (grp2, grp1)}[even_vs_odd_IDs[key]]
```

Comment:
Any specific reason why the keys are "even/odd" and "odd/even" and not simply "even" and "odd"?

Reply:
Yes. We're trying to figure out whether group 1 is odd and group 2 is even or group 2 is odd and group 1 is even.